src/tools/rbd_wnbd/rbd_mapping.cc

```diff
@@ -0,0 +1,286 @@
+#include "common/debug.h"
+#include "common/dout.h"
+#include "common/errno.h"
+
+#include "global/global_init.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd-wnbd: "
+
+#define DISK_STATUS_POLLING_INTERVAL_MS 500
+
+
+int RbdMapping::init()
+{
+  librbd::image_info_t info;
+
+  rados = client_cache.get_client(cfg.entity_name, cfg.cluster_name);
+  if (!rados) {
+    return -EINVAL;
+  }
+
+  int r = rados->ioctx_create(cfg.poolname.c_str(), io_ctx);
+  if (r < 0) {
+    derr << "rbd-wnbd: couldn't create IO context: " << cpp_strerror(r)
+         << ". Pool name: " << cfg.poolname
+         << dendl;
+    return r;
+  }
+
+  io_ctx.set_namespace(cfg.nsname);
+
+  r = rbd.open(io_ctx, image, cfg.imgname.c_str());
+  if (r < 0) {
+    derr << "rbd-wnbd: couldn't open rbd image: " << cpp_strerror(r)
+         << dendl;
+    return r;
+  }
+
+  if (cfg.exclusive) {
+    r = image.lock_acquire(RBD_LOCK_MODE_EXCLUSIVE);
+    if (r < 0) {
+      derr << "rbd-wnbd: failed to acquire exclusive lock: " << cpp_strerror(r)
+           << dendl;
+      return r;
+    }
+  }
+
+  if (!cfg.snapname.empty()) {
+    r = image.snap_set(cfg.snapname.c_str());
+    if (r < 0) {
+      derr << "rbd-wnbd: couldn't use snapshot: " << cpp_strerror(r)
+         << dendl;
+      return r;
+    }
+  }
+
+  r = image.stat(info, sizeof(info));
+  if (r < 0)
+    return r;
+
+  initial_image_size = info.size;
+
+  CephContext* cct = reinterpret_cast<CephContext*>(io_ctx.cct());
+  ceph_assert(cct != nullptr);
+
+  handler = new WnbdHandler(image, cfg.devpath,
+                            info.size / RBD_WNBD_BLKSIZE,
+                            RBD_WNBD_BLKSIZE,
+                            !cfg.snapname.empty() || cfg.readonly,
+                            g_conf().get_val<bool>("rbd_cache"),
+                            cfg.io_req_workers,
+                            cfg.io_reply_workers,
+                            cct->get_admin_socket());
+  return 0;
+}
+
+int RbdMapping::shutdown()
+{
+  std::unique_lock l{shutdown_lock};
+
+  dout(5) << __func__ << ": removing RBD mapping: " << cfg.devpath << dendl;
+
+  int r = 0;
+  if (!cfg.persistent && saved_cfg_to_registry) {
+    dout(5) << __func__ << ": cleaning up non-persistent mapping: "
+            << cfg.devpath << dendl;
+    r = remove_config_from_registry(&cfg);
+    if (r) {
+      derr << __func__ << ": could not clean up non-persistent mapping: "
+           << cfg.devpath << ". Error: " << cpp_strerror(r) << dendl;
+    }
+  }
+
+  if (watch_ctx) {
+    r = image.update_unwatch(watch_handle);
+    if (r < 0) {
+      derr << __func__ << ": update_unwatch failed with error: "
+           << cpp_strerror(r) << dendl;
+    }
+    delete watch_ctx;
+    watch_ctx = nullptr;
+  }
+
+  if (handler) {
+    handler->shutdown();
+    delete handler;
+    handler = nullptr;
+  }
+
+  image.close();
+  io_ctx.close();
+
+  return r;
```

Comment:
With both possible errors logged but one of them getting potentially overridden, do we need to return anything from `shutdown()` at all?  It doesn't look like the return value is checked anywhere, so making this method void would just make that explicit.